_posts/2022-12-19-running-outputs-outside-zonejs-for-angular-performance-optimization.md

```diff
@@ -0,0 +1,288 @@
+        })
+    );
+};
+```
+
+Everything works as expected now: `ChildComponent` does not trigger Change Detection anymore but `ParentComponent` is notified with the new values.
+
+We can see that in this [StackBlitz example](https://stackblitz.com/edit/angular-ivy-komr7r){:target="_blank"}
+The DOM is not rerendered, but `console.log` statements are being made every second and when the user moves the mouse.
+
+## Custom Change Detection
+
+`ChildComponent` does what it needs to do:
+- Run event listeners in the outer zone
+- Notify the parent through `@Output()`s without triggering Change Detection
+
+`ParentComponent` is now responsible of triggering Change Detection and we don't want to trigger it all the time. We only want to trigger it when the mouse of the user moves within the green box.
+
+
+```typescript
+export class ParentComponent {
+    // Inject ChangeDetectorRef because we want to
+    // manually trigger Change Detection with the
+    // detectChanges() function
+    private readonly cdRef= inject(ChangeDetectorRef)
+    val = 0;
+    mouseX = null;
+    mouseY = null;
+
+    public onValChange(v: number): void {
+        // Set the value of val, but don't trigger
+        // Change Detection
+        this.val = v;
+        console.log(v);
+    }
+
+    public onMove(e: { x: number; y: number }): void {
+        // Set the values always
+        this.mouseX = e.x;
+        this.mouseY = e.y;
+        // Only trigger Change Detection when the event
+        // occured inside the green box
+        if(e.x <= 100 && e.y <= 100){
+            this.cdRef.detectChanges();
+        }
+        console.log(e);
+    }
+}
+```
+
+We can find the optimized Stackblitz example [here](https://stackblitz.com/edit/angular-ivy-zkbwrw){:target="_blank"}
+
+## Conclusion
+
+We learned that there are 3 ways of optimizing for Change Detection.
+The first way is by [using the OnPush Change Detection strategy which is described in detail in this article](https://blog.simplified.courses/angular-change-detection-onpush-or-not/){:target="_blank"}. The second one is detaching a component completely from the Change Detector and the third one is to run code in an outer zone.
+We saw how we could achieve this by running `ngZone` its `runOutsideAngular()` function but that was hard to combine with reactive `@Output()`s.
+
+Using Observables on `@Output()`s  gives us the ability to create reactive flows and by creating an `outsideAngular()` operator we were able to create `@Output()`s that notify their parents but do not initiate Change Detection.
+
+While that is probably not something we want to use for everything, it might be beneficial for events that trigger Change Detection too much.
+
+If you like the article, leave me a comment below and if you want to learn more about Change Detection, you can support me by buying this [Angular Change Detection book](https://www.simplified.courses/angular-change-detection-simplified-e-book){:target="_blank"}.
```

Comment:
```suggestion
If you like the article, leave me a comment below and if you want to learn more about Change Detection, you migh be interesteed in buying this [Angular Change Detection book](https://www.simplified.courses/angular-change-detection-simplified-e-book){:target="_blank"}. It will help you understanding and resolveing performance issues in your Angular Enterprise projects.
```